Better multi-site stats fetching using new /me/sites/stats/summary endpoint

mreishus opened the pull request:
## Proposed Changes

* On the existing `/stats/day` and `/stats/month` pages, use a new `/me/sites/stats/summary` endpoint which grabs data for all your sites in one request.

We retain the redux data store's existing structure, but employ a new endpoint for retrieving statistics.

In the present `requestSiteStats()` setup, the system presupposes an existing `siteId`, pre-appending `/sites/${ siteId }/` to API paths. We've introduced an `ALL_SITES_ID` (-1) to work around this limitation, allowing for the use of the new endpoint while preventing potential issues. When `ALL_SITES_ID` is used, it triggers certain new behaviors.

## Testing Instructions

* Visit "/stats/day" and "/stats/month" before and after the patch.
  * After the patch, the data should be the same, but loaded in one XHR, instead of one XHR per account.

mreishus committed:
Try: Using /me/sites/stats/summary endpoint

sgomes commented on client/my-sites/stats/overview.jsx:
```diff
@@ -82,9 +83,16 @@ class StatsOverview extends Component {
 		} );
 
+		const date = moment()
+			.utcOffset( 0 ) // Don't have one site to choose a timezone from...
```
Perhaps we should default to the account timezone, in that case?

mreishus replied:
I'm not finding an account timezone.
Reviewing this, I'm having difficult understanding why a timezone offset is needed. The stats functions used in the endpoint already apply each particular blog's time zone offset to the time periods it generates. 
This calculation on the client is used to determine which day to ask for. But if I ask for the current day in UTC, the endpoint should take that to mean today, then use the offsets to calculate the correct date windows. I feel like this isn't needed at all, but I could be missing something. I set it to UTC for now (without using moment).

sgomes replied:
> I'm not finding an account timezone.

You're right, I mixed up account language with account timezone, which I'm not finding either.

> Reviewing this, I'm having difficult understanding why a timezone offset is needed.

You're right, if the endpoint already handles those calculations, then I'm not sure what's the goal there either 🤔

mreishus replied:
Figured out what was happening with the timezone (there was indeed a problem :) ) and came up with a solution. See lower in the main thread for details.

sgomes commented on client/my-sites/stats/overview.jsx:
```diff
@@ -82,9 +83,16 @@ class StatsOverview extends Component {
 			);
 		} );
 
+		const date = moment()
```
I also really don't like adding yet another usage of `moment` on the client; it's one more call to eventually migrate away from 😞

I don't think we have a reasonable alternative here until Temporal is a thing, though.

mreishus replied:
Changed to `new Date().toISOString().split( 'T' )[ 0 ];` to get the utc date:  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString "The timezone is always UTC"

sgomes replied:
That's great! 👍 If everything is always UTC, then that sounds like a good approach.

sgomes commented on client/my-sites/stats/overview.jsx:
```diff
@@ -82,9 +83,16 @@ class StatsOverview extends Component {
+			.utcOffset( 0 ) // Don't have one site to choose a timezone from...
+			//.utcOffset( Number.isFinite( gmtOffset ) ? gmtOffset : 0 )
+			.format( 'YYYY-MM-DD' );
+		const query = { date, period }; // Makes a new object each render = bad
```
Yeah, the real issue here is `QuerySiteStats` taking its query as an object. That API shape isn't great for a component, precisely because it lends itself to this sort of deopt.

However, the component does have a `componentDidUpdate` method that I added years ago to mitigate the issue, which does a shallow comparison on `query`. So, given that both `date` and `period` are primitives, we should be okay.

mreishus replied:
Oh, nice! I tested it out and it's working perfectly.

mreishus committed:
Move import

mreishus committed:
- Remove moment usage and always use UTC

mreishus committed:
- Query new namespace for endpoint

mreishus commented:
So far it's working in my testing, I see these obvious issues so far:

* On the new endpoint, likes appear as `-` instead of 0
![2023-07-24_10-18](https://github.com/Automattic/wp-calypso/assets/937354/257667ef-87f1-4942-aa10-60ef0b73ba93)
* Calling `dispatch( receiveSiteStats( siteId, singleStatType, query, siteData, Date.now() ) );` in a loop is incredibly slow with 100+ sites; need to make a new action/reducer to update multiple sites at once
* Instead of:
```
	if ( statType === 'allSitesStatsSummary' ) {
			return requestAllSitesStatsSummary( dispatch, query );
		}
```
I should figure out how to use the intended abstraction / organization for new endpoints

sgomes commented:
> I should figure out how to use the intended abstraction / organization for new endpoints

There's a lot of legacy code and deprecated approaches around API calls, so I'm not entirely sure what the current recommendation is. Perhaps @Automattic/team-calypso can provide some guidance there.

mreishus committed:
requestSiteStats: Handle /me/sites/stats/summary endpoint for allSitesStatsSummary

mreishus commented:
> > I should figure out how to use the intended abstraction / organization for new endpoints
> 
> There's a lot of legacy code and deprecated approaches around API calls, so I'm not entirely sure what the current recommendation is. Perhaps @Automattic/team-calypso can provide some guidance there.

I guess I was mostly worried about getting the new stat to work inside the `<QuerySiteStats>` component and `requestSiteStats` action dispatcher, because it's built around the paradigm of requesting for one site and prepending paths with `/sites/${ siteId }/`. I came up with a scheme in b4076ad0c2ec7d5da1b116c2abe7242b9bbe6fb1. Take a look and let me know what you think (still need to make a new receiving action and reducer). For the API itself it's still using `wpcom.req.get` which I think is deprecated.

sgomes commented:
> Take a look and let me know what you think (still need to make a new receiving action and reducer). For the API itself it's still using wpcom.req.get which I think is deprecated.

This is partly why I'd like to see Team Calypso's input here. AFAIK there are a lot of pieces moving towards React Query, and I *think* that's the preferred approach at the moment (although there's still a ton of Redux usage, so I'm not sure). If so, then we're not talking about simply using a different data retrieval mechanism, but an entirely different data handling approach 😕

But maybe it would make sense to keep all of stats in the old paradigm, even if the preferred approach has changed?

mreishus committed:
Make ALL_SITES_STATS_RECEIVE action and reducer

tyxla commented on client/components/data/query-site-stats/index.jsx:
```diff
@@ -36,6 +36,7 @@ class QuerySiteStats extends Component {
 			return;
 		}
 
+		console.log( 'requesting', siteId, statType, query );
```
I think this is a good use case for `debug`. Seems like we're using it for other query components already:

https://github.com/Automattic/wp-calypso/blob/472a10bdefcce5bf500898358a314bdc50ec848e/client/components/data/query-site-purchases/index.jsx#L22

It will allow you to keep the logging statement, but only make it visible to folks who want to see it, without cluttering the console for everyone 😉

tyxla commented on client/my-sites/stats/overview.jsx:
```diff
@@ -82,9 +84,14 @@ class StatsOverview extends Component {
 		} );
 
+		// Generate the current date in yyyy-mm-dd format, in UTC.
+		const date = new Date().toISOString().split( 'T' )[ 0 ];
```
Are we considering the current timezone of the current site when doing this? Could there be any possible issues given the site's potential timezone offset from UTC?

mreishus replied:
The endpoint does calculate each sites' timezone offset and use that when building the queries used to calculate data. That said, I'm still somewhat unclear if something is being missed here. It could be the case.
I think a potential failure could be if UTC and local dates are different, the user sees the text for the UTC date on their screen (but the stats are calculated the same as before). It needs a closer look after we're happy with the data layer.

mreishus replied:
There were issues with the timezone, but I have resolved them now. See lower in the main thread for details.

tyxla commented on client/my-sites/stats/stats-site-overview/index.jsx:
```diff
@@ -3,7 +3,7 @@ import { localize } from 'i18n-calypso';
 import { Component } from 'react';
 import { connect } from 'react-redux';
-import QuerySiteStats from 'calypso/components/data/query-site-stats';
+// import QuerySiteStats from 'calypso/components/data/query-site-stats';
```
Feel free to delet those commented lines altogether instead of commenting them.

tyxla commented on client/state/action-types.ts:
```diff
@@ -946,6 +946,7 @@ export const SITE_SETTINGS_SAVE_FAILURE = 'SITE_SETTINGS_SAVE_FAILURE';
 export const SITE_SETTINGS_SAVE_SUCCESS = 'SITE_SETTINGS_SAVE_SUCCESS';
 export const SITE_SETTINGS_UPDATE = 'SITE_SETTINGS_UPDATE';
 export const SITE_STATS_RECEIVE = 'SITE_STATS_RECEIVE';
+export const ALL_SITES_STATS_RECEIVE = 'ALL_SITES_STATS_RECEIVE';
```
Nit: action types in this file are sorted in alphabetic order. I'd recommend moving this to the top.

tyxla commented on client/state/stats/lists/reducer.js:
```diff
@@ -84,6 +85,27 @@ export const items = withSchemaValidation( itemSchema, ( state = {}, action ) =>
+
+			for ( const siteId in data.stats ) {
+				const siteData = data.stats[ siteId ];
+				const statType = 'statsSummary'; // ??
```
Seems like the `statType` constant can be declared outside of the `for` loop.

tyxla commented on client/state/stats/lists/reducer.js:
```diff
@@ -84,6 +85,27 @@ export const items = withSchemaValidation( itemSchema, ( state = {}, action ) =>
+			const newState = { ...state };
+
+			for ( const siteId in data.stats ) {
+				const siteData = data.stats[ siteId ];
```
`siteData` is used only once; I'd suggest removing the constant and inlining `data.stats[ siteId ]`.

tyxla commented on client/state/stats/lists/reducer.js:
```diff
@@ -84,6 +85,27 @@ export const items = withSchemaValidation( itemSchema, ( state = {}, action ) =>
+				newState[ siteId ] = {
+					...newState[ siteId ],
+					[ statType ]: {
+						...get( newState, [ siteId, statType ] ),
```
Would be perfect if we don't introduce yet another `lodash` usage. I realize it's already been used in this module, but we've aiming to reduce its usage: p4TIVU-9Bf-p2

tyxla commented on client/state/stats/lists/reducer.js:
```diff
@@ -84,6 +85,27 @@ export const items = withSchemaValidation( itemSchema, ( state = {}, action ) =>
+			for ( const siteId in data.stats ) {
+				const siteData = data.stats[ siteId ];
+				const statType = 'statsSummary'; // ??
+				const queryKey = getSerializedStatsQuery( query );
```
Seems like the `queryKey` constant can be declared outside of the `for` loop as well.

tyxla commented on client/state/stats/lists/reducer.js:
```diff
@@ -3,6 +3,7 @@
 import { merge, get } from 'lodash';
 import {
 	SITE_STATS_RECEIVE,
+	ALL_SITES_STATS_RECEIVE,
```
Nit: These imports ideally should also be sorted alphabetically.

tyxla reviewed (commented):
I've done a drive-by review, hope that's helpful.

> > I should figure out how to use the intended abstraction / organization for new endpoints
> 
> There's a lot of legacy code and deprecated approaches around API calls, so I'm not entirely sure what the current recommendation is. Perhaps @Automattic/team-calypso can provide some guidance there.

AFAIK new endpoints are recommended to be `wpcom/v2` endpoints, and they live under the `wp-content/rest-api-plugins` directory, however, this isn't necessarily true for when we're working in a place that uses legacy WPCOM endpoints. If it's easier for code consistency to work with a legacy endpoint, you might as well create a v1 endpoint under `public.api/rest/wpcom-json-endpoints`. If in doubt, I'd recommend consulting with 3hLNG-p2.

sgomes commented:
Thank you for taking a look at this early code, @tyxla! 👍

What's the recommended approach for grabbing data from the API these days, on the Calypso side of things? `wpcom` + Redux, React Query, or something else?

tyxla commented:
> What's the recommended approach for grabbing data from the API these days, on the Calypso side of things? `wpcom` + Redux, React Query, or something else?

It's a complex question because it really depends on how you're going to be utilizing the data, but nowadays React Query (ahem, `@tanstack/query`) is the recommended approach for the most part. We might recommend using Redux only in the cases when there's additional complexity/interactivity involved. When it's about CRUD (which is what most of Calypso's REST API requests are used for nowadays), keeping the SSOT (single source of truth) at the server makes the most sense, and React Query is strong at that, offering all the necessary flexibility and features at a low cost.

You can read more details about the rationale and when to choose what in this recent discussion: p4TIVU-auD-p2

sgomes commented:
Thank you, @tyxla! 🙏

mreishus committed:
some cleanup

mreishus committed:
ALL_SITES_STATS_RECEIVE: pull out stat type

mreishus commented:
Sorry about the debug stuff, I was more at the exploration stage than 'about to commit'.

### Initial Problem

I saw that `/stats/day` issues [one query for each site](https://github.com/Automattic/wp-calypso/blob/dcd0c68178e49af36224212a703dbee3c42ab1bd/client/my-sites/stats/stats-site-overview/index.jsx#L41) and takes a long time to fill out on accounts with a lot (100+ or 1000+) of sites.

Calypso already has an established "stats" data layer in redux, with different stats keys for each type of site, [a helper that takes a stat type + site, makes an API request and dispatches a receiving action](https://github.com/Automattic/wp-calypso/blob/dcd0c68178e49af36224212a703dbee3c42ab1bd/client/state/stats/lists/actions.js#L57), and various other amenities like selectors, etc.

### New Endpoint - Done

We have already shipped a new (v2) endpoint that grabs stats summary for all your sites in one request.
It has room for further optimization, but is working.

### Approach 1 - This PR

Keep the redux data store the same, putting the data in the same place, but use the new endpoint. Because the existing paradigm of `requestSiteStats()` assumes there is always a `siteId`, and does things like prepend `/sites/${ siteId }/` to API paths which we don't want - we have to contort it a bit, inventing a new `ALL_SITES_ID` (-1) to act as a placeholder and trigger some new behavior.

**Pros**:
  - Because we are putting the same data in the same place in the redux store, we are guaranteed anything that reads the data will be compatible with the changes.

**Cons**:
  - More redux :) 
  - The changes to `client/state/stats/lists/actions.js` are awkward because it was originally created with the idea that we would only do one site at a time.

### Approach 2 - Alternate with react-query #79854

Looking closer: Is there anything else in calypso using this static summary data? Looking at the other calls to `getSiteStatsForQuery`, I don't see any other callers passing `'statsSummary'` to the third argument. [This](https://github.com/Automattic/wp-calypso/blob/dcd0c68178e49af36224212a703dbee3c42ab1bd/client/my-sites/stats/stats-site-overview/index.jsx#L90) is the only one. So while the data may be centralized in a redux store here, it seems that it's only used by the `StatsSiteOverview` component.

**Pros**:
  - Using react-query makes the data flow more direct.
  - We don't have to make awkward changes to `client/state/stats/lists/actions.js`.
 
**Cons**:
  - The stats data is now fractured. Most of it is in redux, but the per-site summaries are outside of it and passed via props.
  - Because the component using react-query is a class component, I had to make a quick HOC adapter for react-query which I believe is only designed to be used with hook components. Should probably be extracted but I didn't know where to put it yet.
  - Is it possible the render performance is worse? I know passing objects as props can always be tricky, and redux+selectors could be more optimized than what I'm doing here.

### Next Steps

- Decide on Approach
- Clean up
- Verify the timezone thing closely
- Probably add backend pagination for clients with 1000+ sites - grab in a few requests instead of 1. Would this impact which approach is better?

tyxla commented:
Thanks for the thorough layout of the approaches and where we are @mreishus 🙌 

For the protocol, I'd totally be fine with going with approach 1 and keeping consistency with the current codebase. A potential refactor to React Query could be done separately afterward, and it will be a big project since the stats screens manage lots of data and have plenty of use cases.

Additionally, any performance improvements can be experimented with afterward, too - they don't have to be addressed in this single PR.

mreishus commented:
Ok, there is an issue with the timezones, even though the endpoint is aware of each sites' time offsets. I have a lot of sites across a variety of time zones. It's currently `UTC 2023-07-27 22:45`.  

Calypso (trunk) splits up my sites into two headings: `Stats for July 27, 2023` and `Stats for July 28, 2023`. Most of the sites are July 27, but two of the sites are July 28 - presumably these are in UTC+X timezones where the day has already rolled over.

When calypso (trunk) requests stats, for those two stats, it asks for the 28th (tomorrow from my perspective).

![2023-07-27_17-44](https://github.com/Automattic/wp-calypso/assets/937354/1bcbc21b-eb51-4ee0-98e2-9e85e09b07a1)

However, this patch only asks for the 27th for all sites:

![2023-07-27_17-45](https://github.com/Automattic/wp-calypso/assets/937354/8b88f663-d239-4686-9aff-8c22da3df354)

The selectors ask to pull data out of the "28th" portion of the store, which doesn't have data:

![2023-07-27_17-46](https://github.com/Automattic/wp-calypso/assets/937354/6d9ff560-41dd-471c-9828-4c2cf1133ccd)

Not sure how to fix, but I do have an idea of what's going on now.

mreishus commented:
I tried grouping the sites by date and generating individual stat queries for each group, with each request specifying a list of siteIds:

```diff
diff --git a/client/my-sites/stats/overview.jsx b/client/my-sites/stats/overview.jsx
index b7ce80c0fb..1e91919c6c 100644
--- a/client/my-sites/stats/overview.jsx
+++ b/client/my-sites/stats/overview.jsx
@@ -39,6 +39,32 @@ class StatsOverview extends Component {
 			return site;
 		} );
 
+		// Group sites by date.
+		const sitesByTimezone = sitesSorted.reduce( ( groups, site ) => {
+			const groupKey = site.periodEnd;
+			if ( ! groups[ groupKey ] ) {
+				groups[ groupKey ] = [];
+			}
+			groups[ groupKey ].push( site );
+			return groups;
+		}, {} );
+
+		// Build one <QuerySiteStats /> for each date.
+		const statQueries = Object.entries( sitesByTimezone ).map( ( [ date, sites ] ) => {
+			const query = { date, period };
+			// /me/sites/stats/summary doesn't yet support asking for only some sites.
+			const siteIds = sites.map( ( site ) => site.ID );
+			//siteId={ ALL_SITES_ID /* siteIds */ }
+			return (
+				<QuerySiteStats
+					key={ date }
+					siteId={ siteIds }
+					statType="allSitesStatsSummary"
+					query={ query }
+				/>
+			);
+		} );
+
 		sitesSorted.sort( ( a, b ) => {
 			if ( a.periodEnd > b.periodEnd ) {
 				return 1;
@@ -91,7 +117,7 @@ class StatsOverview extends Component {
 		return (
 			<Main wideLayout>
 				<DocumentHead title={ translate( 'Stats' ) } />
-				<QuerySiteStats siteId={ ALL_SITES_ID } statType="allSitesStatsSummary" query={ query } />
+				{ statQueries }
 				<PageViewTracker
 					path={ `/stats/${ period }` }
 					title={ `Stats > ${ titlecase( period ) }` }
diff --git a/client/state/stats/lists/actions.js b/client/state/stats/lists/actions.js
index e5d5f6e7d5..2a1124fc58 100644
--- a/client/state/stats/lists/actions.js
+++ b/client/state/stats/lists/actions.js
@@ -81,17 +81,21 @@ export function requestSiteStats( siteId, statType, query ) {
 
 		let subpath;
 		let apiNamespace;
+		const isMultipleSiteIds = Array.isArray( siteId );
 		if ( wpcomV1Endpoints.hasOwnProperty( statType ) ) {
 			subpath = wpcomV1Endpoints[ statType ];
 		} else if ( wpcomV2Endpoints.hasOwnProperty( statType ) ) {
 			subpath = wpcomV2Endpoints[ statType ];
 			apiNamespace = 'wpcom/v2';
-		} else if ( siteId === ALL_SITES_ID && wpcomV2AllSitesEndpoints.hasOwnProperty( statType ) ) {
+		} else if (
+			( isMultipleSiteIds || siteId === ALL_SITES_ID ) &&
+			wpcomV2AllSitesEndpoints.hasOwnProperty( statType )
+		) {
 			subpath = wpcomV2AllSitesEndpoints[ statType ];
 			apiNamespace = 'wpcom/v2';
 		}
 
-		const options = ( () => {
+		let options = ( () => {
 			switch ( statType ) {
 				case 'statsVideo':
 					return query.postId;
@@ -114,10 +118,17 @@ export function requestSiteStats( siteId, statType, query ) {
 			}
 		} )();
 
+		if ( isMultipleSiteIds ) {
+			options = { ...options, siteId };
+		}
+
 		let requestStats;
 
 		if ( subpath ) {
-			const path = siteId === ALL_SITES_ID ? `${ subpath }` : `/sites/${ siteId }/${ subpath }`;
+			const path =
+				isMultipleSiteIds || siteId === ALL_SITES_ID
+					? `${ subpath }`
+					: `/sites/${ siteId }/${ subpath }`;
 			requestStats = wpcom.req.get(
 				{
 					path,
```

However, the URLs generated look quite silly/excessive if you have 100+ sites, and some agencies have 1000+ sites. Maybe we need the backend to do the work of figuring out date+site combinations?

tyxla commented:
Is the timezone issue present in production, or does it get exposed when you use the new endpoint? I guess if it's not present, perhaps we could compare and address it the same way it was handled before.

Otherwise, it does make sense to be able to address the timezone discrepancies on the backend; otherwise we'd have to update a whole ton of components and we'd still not have a guarantee that we haven't missed any edge case.

mreishus committed:
Pass empty date to summary endpoint and create new query keys for each site

mreishus commented:
The timezone issue is not present on production. It's more of a function of the granularity of requests doesn't match exactly with the new endpoint. However, I've found a solution.

### Example - Trunk

Let's say it's 2023-07-31, 22:00 UTC and I have 3 different sites in different timezones:
```
ABC1 UTC-5  Local Today is 2023-07-31
ABC2 UTC    Local Today is 2023-07-31
ABC3 UTC+5  Local Today is 2023-08-01
```
Calypso will do requests of:
```
 /sites/ABC1/stats/summary?date=2023-07-31
 /sites/ABC2/stats/summary?date=2023-07-31
 /sites/ABC3/stats/summary?date=2023-08-31
```
Now, if it's earlier in the same UTC day and instead 2023-07-31, 02:00 UTC
We have:
```
ABC1 UTC-5  Local Today is 2023-07-30
ABC2 UTC    Local Today is 2023-07-31
ABC3 UTC+5  Local Today is 2023-07-31
```
And calypso (trunk) will do:
```
 /sites/ABC1/stats/summary?date=2023-07-30
 /sites/ABC2/stats/summary?date=2023-07-31
 /sites/ABC3/stats/summary?date=2023-07-31
```

So Calypso is asking for different days on different sites, and different behavior can happen even on the same UTC day (we see different requests made early on 2023-07-31 UTC and late on 2023-07-31).

### Example - This branch (before fix)

All this branch could do is request: `/me/sites/stats/summary?date=2023-07-31`, which means it would not always fetch 'today' for sites in different time zones. So if it's 2023-07-31, 22:00 UTC Calypso will want to display results keyed for 2023-08-01 for the site ABC since it's UTC+5, but it won't be there in the redux store:

![2023-07-27_17-46](https://github.com/Automattic/wp-calypso/assets/937354/303224bd-4e9e-4359-813f-1f3cc1bf5661)

### Example - This branch (after fix)

The new endpoint, if given a blank, empty or missing date, will use each blog's current date right now in local time. This is essentially moving the logic over from calypso to the endpoint.

Let's re-examine the same example. If it's 2023-07-31, 22:00 UTC, and I request:
```
/me/sites/stats/summary?date=
```
I'll get back
```
ABC1 - UTC-5 - Stats for 2023-07-31 with "date": 2023-07-31 in the object returned
ABC2 - UTC - Stats for 2023-07-31 with "date": 2023-07-31 in the object returned
ABC3 - UTC+5 - Stats for 2023-08-01 with "date":2023-08-01 in the object returned
```
And if it's 2023-07-31 02:00 UTC and I request:
```
/me/sites/stats/summary?date=
```
I'll get back
```
ABC1 - UTC-5 - Stats for 2023-07-30 with "date": 2023-07-30 in the object returned
ABC2 - UTC - Stats for 2023-07-31 with "date": 2023-07-31 in the object returned
ABC3 - UTC+5 - Stats for 2023-07-31 with "date":2023-07-31 in the object returned
```

This is matching the pattern of requests that Calypso makes on trunk, and puts the data in the same place in redux after I make a small change to the new reducer added in this PR ( 5ea91aa088 ).  The endpoint change is pretty simple too: D117597-code.

mreishus commented:
Ok, this is ready for review. `/stats/day` and `/stats/month` should be showing the exact same data as on trunk, no matter what timezones your sites have or what time of day it is, but fetched over only one request instead of one per site.

tyxla reviewed (commented):
Thanks for the great work @mreishus!

I haven't tested again, but code-wise things are looking good to me and my previous concerns have been addressed.

sgomes reviewed (approved):
The code LGTM, with no red flags, and the live branch tests well for me 👍

Great work, @mreishus!

mreishus committed:
requestSiteStats: let all sites request do pagination

mreishus committed:
extract requestAllSiteStats()

mreishus commented:
> Thanks for the great work @mreishus!
> 
> I haven't tested again, but code-wise things are looking good to me and my previous concerns have been addressed.

> The code LGTM, with no red flags, and the live branch tests well for me +1
> 
> Great work, @mreishus!

Thank you for the reviews!
After some more consideration, I don't think it was production ready. My main concern was the agency accounts. While my user with ~200 accounts works great here, I think an agency account with 1,500 sites or 2,500 sites could run into some problems with the endpoint timing out and possibly causing OOM errors on our end.

I've done the following:
* D118198-code: Added pagination support to the endpoint:
* fbc202d008: Changed the `requestSiteStats()` function to recursively issue more fetches to page through all available sites
* 78ed12572b12: Split out `requestAllSiteStats()` from `requestSiteStats()`. After I added the pagination code, I felt that I really diluted the original `requestSiteStats()` with a bunch of code specific to this one endpoint, so I wanted to pull that out.

I haven't pushed D118198-code yet: will do Monday. I just couldn't bring myself to do it on a Friday evening.

mreishus committed:
undo unneeded change

mreishus commented:
> Interesting, do we have any/many of those? If we do, we might have to spend some time testing and optimizing many other multisite parts of Calypso.

I'm not sure how many there are, but these large agency accounts definitely exist: 
- p55Cj4-2xe-p2
- p1689177278736889-slack-C4GAQ900P

@JoshuaGoode brought this one to my attention and may be aware of other pain points for calypso with large accounts.

mreishus commented:
Ok, all the backend changes are set up and in production now. Planning on deploying this tomorrow.
